app/src/main/java/org/schabi/newpipe/player/resolver/ExtractorDataSourceResolver.java

```diff
@@ -0,0 +1,321 @@
+package org.schabi.newpipe.player.resolver;
+
+import android.net.Uri;
+import android.util.Base64;
+import android.util.Log;
+
+import com.google.android.exoplayer2.upstream.DataSpec;
+import com.google.android.exoplayer2.upstream.ResolvingDataSource;
+
+import org.schabi.newpipe.extractor.stream.AudioStream;
+import org.schabi.newpipe.extractor.stream.Stream;
+import org.schabi.newpipe.extractor.stream.StreamInfo;
+import org.schabi.newpipe.extractor.stream.SubtitlesStream;
+import org.schabi.newpipe.extractor.stream.VideoStream;
+import org.schabi.newpipe.player.Player;
+import org.schabi.newpipe.util.ExtractorHelper;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+public final class ExtractorDataSourceResolver implements ResolvingDataSource.Resolver {
```

Comment:
I'm not sure if 3 interfaces and 4 classes inside that class are a good idea. Maybe create separate classes?